File: p785m/is_bipartite.js
/**
 * @param {number[][]} graph
 * @return {boolean}
 */
var isBipartite = function (graph) {
  const party = Array.from({ length: graph.length });
  // indicate which group, T or F a node belongs to
  // if undefined means not visited yet
  for (let i in graph) {
    if (party[i] !== undefined) continue;
    // BFS traverse
    party[i] = true;
    let bfsQueue = [i];
    while (bfsQueue.length > 0) {
      let u = bfsQueue.shift();
      for (let v of graph[i]) {
        if (party[v] !== undefined) {
          if (party[v] === party[u]) return false;
        } else {
          bfsQueue.push(v);
          party[v] = !party[u];
        }
      }
    }
  }
  return true;
};
// TEST
[
  [
    [
      [1, 2, 3],
      [0, 2],
      [0, 1, 3],
      [0, 2],
    ],
    false,
  ],
  [
    [
      [1, 3],
      [0, 2],
      [1, 3],
      [0, 2],
    ],
    true,
  ],
  [[[3], [2, 4], [1], [0, 4], [1, 3]], true],
  [[[1], [0, 3], [3], [1, 2]], true],
  [[[1], [0]], true],
  [[[1, 2], [0], [0]], true],
  [
    [
      [1, 2],
      [0, 2],
      [0, 1],
    ],
    false,
  ],
  [
    [
      [],
      [2, 4, 6],
      [1, 4, 8, 9],
      [7, 8],
      [1, 2, 8, 9],
      [6, 9],
      [1, 5, 7, 8, 9],
      [3, 6, 9],
      [2, 3, 4, 6, 9],
      [2, 4, 5, 6, 7, 8],
    ],
    false,
  ],
].forEach(([graph, expected]) => {
  const actual = isBipartite(graph);
  console.log('Is graph', graph, 'bipartite? ->', actual);
  console.assert(actual === expected);
});

File: p785m/is_bipartite.py
from typing import List


class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        party = [None] * len(graph)
        for i in range(len(graph)):
            if party[i] is not None:
                continue
            party[i] = True
            queue = [i]
            while queue:
                u = queue.pop(0)
                for v in graph[u]:
                    if party[v] is None:
                        party[v] = not party[u]
                        queue.append(v)
                    else:
                        if party[v] == party[u]:
                            return False
        return True


# TESTS
for graph, expected in [
    [[[1, 2, 3], [0, 2], [0, 1, 3], [0, 2]], False],
    [[[1, 3], [0, 2], [1, 3], [0, 2]], True],
    [[[3], [2, 4], [1], [0, 4], [1, 3]], True],
    [[[1], [0, 3], [3], [1, 2]], True],
    [[[1], [0]], True],
    [[[1, 2], [0], [0]], True],
    [[[1, 2], [0, 2], [0, 1]], False],
    [
        [
            [],
            [2, 4, 6],
            [1, 4, 8, 9],
            [7, 8],
            [1, 2, 8, 9],
            [6, 9],
            [1, 5, 7, 8, 9],
            [3, 6, 9],
            [2, 3, 4, 6, 9],
            [2, 4, 5, 6, 7, 8],
        ],
        False,
    ],
]:
    sol = Solution()
    actual = sol.isBipartite(graph)
    print("Is graph", graph, "bipartite? ->", actual)
    assert actual == expected
